scipy/ndimage/src/nd_image.c

```diff
@@ -934,6 +934,227 @@ static PyObject *Py_FindObjects(PyObject *obj, PyObject *args)
     }
 }
 
+/*
+   Implement the ndimage.value_indices() function.
+   Makes 3 passes through the array data. We use ndimage's NI_Iterator to
+   iterate through all elements of the input array.
+
+   To support all numpy datatypes, we define several macros which take a
+   parameter for the array datatype in question. By using these, all comparisons
+   are carried out using the same datatype as the input array.
+*/
+#define VALUEINDICES_MINVAL(valType) (*((valType *)PyArray_GETPTR1(minMaxArr, 0)))
+#define VALUEINDICES_MAXVAL(valType) (*((valType *)PyArray_GETPTR1(minMaxArr, 1)))
+#define VALUEINDICES_IGNOREVAL(valType) (*((valType *)PyArray_GETPTR1(ignorevalArr, 0)))
+#define CASE_VALUEINDICES_SET_MINMAX(valType) {\
+    valType val = *((valType*)arrData); \
+    if (nullIsNone || (val != VALUEINDICES_IGNOREVAL(valType))) {\
+        if (minMaxUnset) { \
+            VALUEINDICES_MINVAL(valType) = val;  \
+            VALUEINDICES_MAXVAL(valType) = val; \
+            minMaxUnset = 0; \
+        } else { \
+            if (val < VALUEINDICES_MINVAL(valType)) VALUEINDICES_MINVAL(valType) = val; \
+            if (val > VALUEINDICES_MAXVAL(valType)) VALUEINDICES_MAXVAL(valType) = val; \
+        } \
+    }\
+}
+#define CASE_VALUEINDICES_MAKEHISTOGRAM(valType) {\
+    numPossibleVals = (VALUEINDICES_MAXVAL(valType) - VALUEINDICES_MINVAL(valType) + 1); \
+    hist = (npy_intp *)calloc(numPossibleVals, sizeof(npy_intp)); \
+    if (hist != NULL) { \
+        NI_InitPointIterator(arr, &ndiIter); \
+        arrData = (char *)PyArray_DATA(arr); \
+        for (iterIndex=0; iterIndex<arrSize; iterIndex++) { \
+            valType val = *((valType*)arrData); \
+            if (nullIsNone || (val != VALUEINDICES_IGNOREVAL(valType))) { \
```

Comment:
What is `nullIsNone` used for here?

Reply:
Because it is possible to give None as the ignore_value, every time we check for an array value being equal to the ignore_value, we must first check whether the ignore_value was None. Using None seemed the most robust (and most Pythonic) way to choose not to ignore any of the values, without making further assumptions.